### PR TITLE
feat: add useMap and useSet

### PR DESCRIPTION
**Pull request by jquense**
w/d/t about these? useful for when you want a set/map value in state but don't want to keep remaking it

cc @manon-pilaud

### CODE REVIEW

File: src/useForceUpdate.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { useCallback, useState } from 'react'
+import { useCallback, useState, useReducer } from 'react'
```
Comment (TrySound) [inline src/useForceUpdate.ts]: useCallback and useState are unused

File: src/useMap.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import useForceUpdate from './useForceUpdate'
+import { useRef } from 'react'
+import useWillUnmount from './useWillUnmount'
+
+export class ObservableMap<K, V> extends Map<K, V> {
+  private readonly listeners = new Set<Function>()
```
Comment (taion) [inline src/useMap.ts]: You can probably just take a single, required listener in the ctor. No need to track multiple listeners or deal with empty

File: src/useMap.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  }
+
+  unobserve() {
+    this.listeners.clear()
+  }
+
+  private emit() {
+    // the ctor init will call set() so the listener Set isn't set up yet
+    if (!this.listeners) return
```
Comment (taion) [inline src/useMap.ts]: not that it matters, but the `Set` will always be truth-y

File: src/useMap.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  const forceUpdate = useForceUpdate()
+  const set = useRef<ObservableMap<K, V>>()
+  if (!set.current) {
+    set.current = new ObservableMap<K, V>(init as any)
+    set.current.observe(forceUpdate)
+  }
+
+  useWillUnmount(() => {
+    if (set.current) {
+      set.current.unobserve()
```
Comment (taion) [inline src/useMap.ts]: Is this really necessary? If the set still gets updated after unmount, then this seems like it'd be a component bug

File: src/useMap.ts (hunk 3 above)
Comment (jquense) [reply src/useMap.ts]: it won't! super() in the ctor fires before you can assign the Set to `this.listeners`

File: src/useMap.ts (hunk 4 above)
Comment (jquense) [reply src/useMap.ts]: not strictly, but seemed easy enough. I guess a console warning tho is probably better for surfacing a potential bug

File: src/useMap.ts (hunk 2 above)
Comment (jquense) [reply src/useMap.ts]: yeah, tbh i didn't want to type the ctor, then we need to keep it up to date with whatever Set/Map take, which depends on on the `lib` setup in your tsconfig

Comment (taion) [reply src/useMap.ts]: still don't need to support multiple listeners :p

**Pull request merged**